### PR TITLE
8306471: Add virtual threads support to JDWP ThreadReference.Stop and JDI ThreadReference.stop()

### PR DESCRIPTION
**Pull request by plummercj**
[JDK-8306034](https://bugs.openjdk.org/browse/JDK-8306034) (#13546) added some virtual thread support to JVMTI StopThread. This allows JDWP ThreadReference.Stop and JDI ThreadReference.stop() to have the same level support for virtual threads.

For your reference, here are the JVMTI spec changes that this pr is using as the basis for the JDWP and JDI spec updates.

https://github.com/openjdk/jdk/pull/13546/files#diff-cab9797fde25f344784e4473778cd35bfd9e19ba36f82c11b1a0674636958339

Mostly this is a spec update for JDWP and JDI, but there are also some minor changes needed to the ThreadReference.stop() handling of JDWP errors, and throwing the appropriate exceptions. Also some minor cleanup in jdb. The debug agent doesn't need changes since JVMTI errors are just passed through as the corresponding JDWP errors, and the code for this is already in place. These errors are not new nor need any special handling.

Our existing testing for ThreadReference.stop() is fairly weak:

- nsk/jdb/kill/kill001, which tests stop() when the thread is suspended at a breakpoint. It will get problem listed by [JDK-8306034](https://bugs.openjdk.org/browse/JDK-8306034). I have fixes for it already working and will push it separately.
- nsk/jdi/stop/stop001, which is problem listed and only tests when the thread is blocked in Object.wait()
- nsk/jdi/stop/stop002, which only tests that throwing an invalid exception fails properly

I decided to take stop002 and make it a more thorough test of ThreadReference.stop(). See the comment at the top of the test for a list of what is tested. As for reviewing this test, it's probably best to look at it as a completely new test rather than looking at diffs since so much has been changed and added.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8306472](https://bugs.openjdk.org/browse/JDK-8306472) to be approved

### Issues
 * [JDK-8306471](https://bugs.openjdk.org/browse/JDK-8306471): Add virtual threads support to JDWP ThreadReference.Stop and JDI ThreadReference.stop()
 * [JDK-8306472](https://bugs.openjdk.org/browse/JDK-8306472): Add virtual threads support to JDWP ThreadReference.Stop and JDI ThreadReference.stop() (**CSR**)

### Reviewers
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13548/head:pull/13548` \
`$ git checkout pull/13548`

Update a local copy of the PR: \
`$ git checkout pull/13548` \
`$ git pull https://git.openjdk.org/jdk.git pull/13548/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13548`

View PR using the GUI difftool: \
`$ git pr show -t 13548`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13548.diff">https://git.openjdk.org/jdk/pull/13548.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13548#issuecomment-1522289832)

### CODE REVIEW

File: test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/stop/stop002.java
Comment (turbanoff) [inline test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/stop/stop002.java:169]: nit
```suggestion
            } catch (Exception ue) {
```

File: test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/stop/stop002.java
Comment (turbanoff) [inline test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/stop/stop002.java:192]: nit
```suggestion
            } catch (Exception ue) {
```

File: test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/stop/stop002.java
Comment (turbanoff) [inline test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/stop/stop002.java:223]:
```suggestion
            } catch (Throwable ue) {
```

File: test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/stop/stop002.java
Comment (turbanoff) [inline test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/stop/stop002.java:255]:
```suggestion
            } catch (Throwable ue) {
```

File: test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/stop/stop002.java
Comment (turbanoff) [inline test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/stop/stop002.java:274]:
```suggestion
            if (objRef != null && stopLoop1 != null && stopLoop2 != null) {
```

File: src/java.se/share/data/jdwp/jdwp.spec
Comment (sspitsyn) [inline src/java.se/share/data/jdwp/jdwp.spec:2024]: Should it be aligned with JVMTI and some other places in your fix and say
 "from the current frame" instead of "from this frame"?

File: src/jdk.jdi/share/classes/com/sun/jdi/ThreadReference.java
Comment (sspitsyn) [inline src/jdk.jdi/share/classes/com/sun/jdi/ThreadReference.java:132]: The same comment as for the jdwp spec:
Should it be aligned with JVMTI and some other places in your fix and say
 "from the current frame" instead of "from this frame"?

File: test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/stop/stop002.java
Comment (sspitsyn) [inline test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/stop/stop002.java:39]: Typo: "I consists" => "It consists".

File: test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/stop/stop002t.java
Comment (sspitsyn) [inline test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/stop/stop002t.java:78]: Where is a similar comment for TEST #1 ?

File: src/java.se/share/data/jdwp/jdwp.spec
Comment (plummercj) [reply src/java.se/share/data/jdwp/jdwp.spec:2024]: I'm waiting for your JVMTI PR to finish review. I don't want to have to change this more than once.

File: src/jdk.jdi/share/classes/com/sun/jdi/ThreadReference.java
Comment (plummercj) [reply src/jdk.jdi/share/classes/com/sun/jdi/ThreadReference.java:132]: I'm waiting for your JVMTI PR to finish review. I don't want to have to change this more than once.

File: test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/stop/stop002.java
Comment (sspitsyn) [inline test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/stop/stop002.java:63]: This comment is a little bit confusing.
Should it be one line up or it is intentionally placed before `DEBUGGEE_THROWABLE_VAR`?

File: src/jdk.jdi/share/classes/com/sun/tools/jdi/ThreadReferenceImpl.java
Comment (plummercj) [reply src/jdk.jdi/share/classes/com/sun/tools/jdi/ThreadReferenceImpl.java:279]: The implementation tends to only include an exception message when it helps to clarify the reason for the exception. The spec only gives one possible reason for this exception, so no clarification should be needed. Plus it would be hard to meaningfully convey this reason in a short message. Here's what the spec says:

     * @throws OpaqueFrameException if the thread is a suspended
     * virtual thread and the implementation was unable to throw an
     * asynchronous exception from this frame

File: test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/stop/stop002t.java
Comment (plummercj) [reply test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/stop/stop002t.java:78]: Test #1 does not involve the debuggee since it is suppose to fail on the JDI side with an exception.

File: test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/stop/stop002.java
Comment (sspitsyn) [inline test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/stop/stop002.java:300]: Do we need a sleep at each iteration?

File: test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/stop/stop002.java
Comment (plummercj) [reply test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/stop/stop002.java:63]: It should be "throwable", not "non-throwable"

File: test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/stop/stop002.java
Comment (sspitsyn) [inline test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/stop/stop002.java:148]: The same suggestion about refactoring each subtest case into a separate method.
There is some overhead in doing this, so it is up to you but worth to consider.

File: test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/stop/stop002.java
Comment (plummercj) [reply test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/stop/stop002.java:300]: I intentionally did not do one since `getValue()` is fairly slow and involves sending and receiving a JDWP packet. It didn't seem worth the extra noise of a sleep call and the needed exception handling.

File: src/jdk.jdi/share/classes/com/sun/jdi/ThreadReference.java
Comment (sspitsyn) [reply src/jdk.jdi/share/classes/com/sun/jdi/ThreadReference.java:132]: Okay.

File: src/jdk.jdi/share/classes/com/sun/tools/jdi/ThreadReferenceImpl.java
Comment (sspitsyn) [reply src/jdk.jdi/share/classes/com/sun/tools/jdi/ThreadReferenceImpl.java:279]: Okay, thanks.

File: test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/stop/stop002.java
Comment (sspitsyn) [reply test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/stop/stop002.java:300]: Okay, thanks.

File: test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/stop/stop002t.java
Comment (sspitsyn) [reply test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/stop/stop002t.java:78]: > Test https://github.com/openjdk/jdk/pull/1 does not involve the debuggee
> since it is suppose to fail on the JDI side with an exception.

Still some comment is needed to explain this.
Otherwise, it is confusing that the subtests start from the TEST #2. :)

Comment (plummercj) [reply test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/stop/stop002t.java:85]: Fixed.

File: src/java.se/share/data/jdwp/jdwp.spec
Comment (plummercj) [reply src/java.se/share/data/jdwp/jdwp.spec]: I've done the update.

File: src/jdk.jdi/share/classes/com/sun/jdi/ThreadReference.java
Comment (plummercj) [reply src/jdk.jdi/share/classes/com/sun/jdi/ThreadReference.java]: I've don the updae

**Comment by plummercj** [comment]
This PR has been updated after the JVMTI push of the JVMTI StopThread pr. Testing is in progress.

File: test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/stop/stop002.java
Comment (sspitsyn) [inline test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/stop/stop002.java:154]: It is confusing that the log.complain at line 152 expects `IllegalArgumentException` but the really expected exception is `InvalidTypeException`. Is it intentional? If so then it is better to clarify it in the comment.

**Review by sspitsyn — APPROVED** [review]
This looks good to me.
I've posted one question/nit though.

Thanks,
Serguei

File: test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/stop/stop002.java
Comment (plummercj) [reply test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/stop/stop002.java:154]: Yes, this looks like a pre-existing bug. I'll cleanup the message.

Comment (sspitsyn) [reply test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/stop/stop002.java:154]: Thank you for the update!

**Review by sspitsyn — APPROVED** [review]

File: src/java.se/share/data/jdwp/jdwp.spec
Comment (AlanBateman) [inline src/java.se/share/data/jdwp/jdwp.spec:2025]: This is okay but could be a bit clearer with "from the thread's current frame".

File: src/jdk.jdi/share/classes/com/sun/jdi/ThreadReference.java
Comment (AlanBateman) [inline src/jdk.jdi/share/classes/com/sun/jdi/ThreadReference.java:132]: I think "from the thread's current frame" might be a bit clearer here, more so than the JDWP spec because the current frame could be read as the current JDI (if you see what I mean). What you have is okay too, but like the JDWP spec, could be a bit clearer.

**Review by AlanBateman — APPROVED** [review]
I've looked at the spec + implementation changes, didn't have time to study the tests closely. Only comment is "current frame".

File: src/jdk.jdi/share/classes/com/sun/jdi/ThreadReference.java
Comment (plummercj) [reply src/jdk.jdi/share/classes/com/sun/jdi/ThreadReference.java:132]: Ok, but JVMTI doesn't use the wording you suggest, and it seems it is in the same situation.

```
        <error id="JVMTI_ERROR_OPAQUE_FRAME">
          The thread is a suspended virtual thread and the implementation was unable
          to throw an asynchronous exception from the current frame.
```

Comment (AlanBateman) [reply src/jdk.jdi/share/classes/com/sun/jdi/ThreadReference.java:132]: > Ok, but JVMTI doesn't use the wording you suggest, and it seems it is in the same situation.

I don't disagree, it's just a bit more pronounced with JDI because the caller is in different VM. It's the "current frame" of the Thread in the target VM rather than the current VM.  I'll leave it to you as to whether you want to re-visit this wording or not.

File: src/java.se/share/data/jdwp/jdwp.spec
Comment (plummercj) [reply src/java.se/share/data/jdwp/jdwp.spec]: I've updated the text.

File: src/jdk.jdi/share/classes/com/sun/jdi/ThreadReference.java
Comment (plummercj) [reply src/jdk.jdi/share/classes/com/sun/jdi/ThreadReference.java]: I've updated the text.

**Review by AlanBateman — APPROVED** [review]

**Review by sspitsyn — APPROVED** [review]

**Comment by plummercj** [comment]
Thanks Serguei and Alan!

/integrate

**Pull request closed without merging**
